File: runtime/go/trace.go
package runtime

import (
	"context"
	"crypto/rand"
	"encoding/hex"
)

// Trace is the W3C Trace Context a frame carries: the two members verbatim,
// neither of them read by the runtime beyond the form the decoder holds a
// traceparent to. The runtime imports no tracing library; one binds to it as a
// Propagator and nowhere else.
type Trace struct{ Parent, State string }

// Propagator moves a trace between a context and a frame. Extract places an
// incoming frame's trace in the context its handler runs under; Inject says
// what an outgoing frame carries — a child of the trace the context holds, or
// a new trace where it holds none. Both members reach it verbatim, and a peer
// sends what it returns as it returns it.
type Propagator interface {
	Extract(ctx context.Context, trace Trace) context.Context
	Inject(ctx context.Context) Trace
}

// DefaultPropagator is what a peer with no Options.Propagator propagates by: it
// keeps an incoming trace verbatim and mints an outgoing one's ids itself, so
// that frames correlate across hops with no tracing library installed.
var DefaultPropagator Propagator = w3cPropagator{}

type traceKey struct{}

// TraceOf is the trace DefaultPropagator placed in ctx, and whether one is
// there. A propagator of another making keeps its trace where it chooses.
func TraceOf(ctx context.Context) (Trace, bool) {
	trace, ok := ctx.Value(traceKey{}).(Trace)
	return trace, ok
}

// w3cPropagator carries the trace under a key of its own and mints ids in the
// one form W3C Trace Context gives them: version 00, sampled, a tracestate
// forwarded exactly as it arrived.
type w3cPropagator struct{}

// Extract keeps a trace whose traceparent the decoder has already held to its
// form. A frame carrying none leaves the context as it was, so that what is
// sent from there begins a trace rather than continuing one that is not there.
func (w3cPropagator) Extract(ctx context.Context, trace Trace) context.Context {
	if trace.Parent == "" {
		return ctx
	}
	return context.WithValue(ctx, traceKey{}, trace)
}

// Inject mints a child of the context's trace — its trace id and its flags, a
// span id of this frame's own — or a new trace where the context carries none.
func (w3cPropagator) Inject(ctx context.Context) Trace {
	parent, ok := TraceOf(ctx)
	if !ok || !validTraceparent(parent.Parent) {
		return Trace{Parent: "00-" + randomID(16) + "-" + randomID(8) + "-01"}
	}
	return Trace{Parent: parent.Parent[:36] + randomID(8) + parent.Parent[52:], State: parent.State}
}

// randomID is n random bytes in lower-case hexadecimal. crypto/rand fills the
// buffer or does not return, so a trace id is never the zero one by accident.
func randomID(n int) string {
	id := make([]byte, n)
	rand.Read(id)
	return hex.EncodeToString(id)
}

File: runtime/ts/src/trace.ts
import type { RequestContext } from './peer.ts';

/**
 * W3C Trace Context as the wire carries it: the two members verbatim, nothing
 * invented and nothing normalised. An absent member is the empty string, which
 * a frame never carries; a traceparent is the one of the two the peer holds to
 * a form, and the propagator decides what either of them means.
 */
export interface Trace {
  traceparent: string;
  tracestate?: string;
}

/**
 * Where an incoming trace goes and what an outgoing frame carries. The runtime
 * imports no tracing library: an adapter for one replaces this hook, and the
 * default below correlates without it.
 */
export interface Propagator {
  /** Places an incoming frame's trace on the context its handler runs with. */
  extract(context: RequestContext, trace: Trace | undefined): void;
  /** What an outgoing frame carries: a child of the context's trace, or a new trace. */
  inject(context: RequestContext | undefined): Trace;
}

/** `version-traceid-spanid-flags`, the one form the profile accepts, in its parts. */
const TRACEPARENT = /^([0-9a-f]{2})-([0-9a-f]{32})-[0-9a-f]{16}-([0-9a-f]{2})$/;

/**
 * Correlation with no tracing library installed: a new trace is a random
 * 16-byte trace id and 8-byte span id, sampled; a child keeps its parent's
 * version, trace id and flags and mints a span id of its own, and carries the
 * parent's tracestate verbatim.
 */
export const defaultPropagator: Propagator = {
  extract(context, trace) {
    // A tracestate that arrives without a traceparent continues no trace, as
    // the specification says; the response to its frame still carries it back.
    if (trace?.traceparent) context.trace = trace;
  },
  inject(context) {
    const trace = context?.trace;
    if (trace) {
      const parent = TRACEPARENT.exec(trace.traceparent);
      if (parent) {
        const child: Trace = { traceparent: `${parent[1]}-${parent[2]}-${randomHex(8)}-${parent[3]}` };
        if (trace.tracestate !== undefined) child.tracestate = trace.tracestate;
        return child;
      }
    }
    return { traceparent: `00-${randomHex(16)}-${randomHex(8)}-01` };
  },
};

/** The members an incoming frame carries, verbatim: what a propagator extracts. */
export function traceOf(frame: Record<string, unknown>): Trace | undefined {
  const traceparent = typeof frame.traceparent === 'string' ? frame.traceparent : '';
  const tracestate = typeof frame.tracestate === 'string' ? frame.tracestate : '';
  if (!traceparent && !tracestate) return undefined;
  return tracestate ? { traceparent, tracestate } : { traceparent };
}

/** Stamps onto an outgoing frame what a propagator minted; an empty member is none. */
export function traced(envelope: Record<string, unknown>, trace: Trace | undefined): Record<string, unknown> {
  if (trace?.traceparent) envelope.traceparent = trace.traceparent;
  if (trace?.tracestate) envelope.tracestate = trace.tracestate;
  return envelope;
}

/** Web Crypto is the only source; the runtime takes no dependency for it. */
function randomHex(bytes: number): string {
  return Array.from(crypto.getRandomValues(new Uint8Array(bytes)), byte => byte.toString(16).padStart(2, '0')).join('');
}
